lib/clustering/mean_shift.cpp

```diff
@@ -6,39 +6,113 @@
 
 
 #include "mean_shift.h"
-#include <stdlib.h>
-#include <time.h>
+#include <boost/log/trivial.hpp>
+#include <Eigen/Dense>
+
 namespace MouseTrack {
 
-MeanShift::MeanShift(int k) : _k(k) {
-    //empty
+MeanShift::MeanShift(double window_size) : _window_size(window_size) {
+	//empty
 }
 
 std::vector<Cluster> MeanShift::operator()(const PointCloud& cloud) const {
+	BOOST_LOG_TRIVIAL(trace) << "MeanShift algorithm started";
+	// Covert point cloud to Eigen vectors
+	std::vector<Eigen::VectorXd> points(cloud.size());
+	for (int i=0; i<cloud.size(); i++) {
+		points[i] = cloud[i].eigenVec();
+	}
+
+
+	// Initialize Clusters. Initially, every point has its own cluster.
+	std::vector<Cluster> clusters(cloud.size());
+	for (PointIndex i=0; i<cloud.size(); i++) {
+		 clusters[i].points().push_back(i);
+
+	}
+
+	// Initialize some stuff used in the MeanShift loop
+	std::vector<Eigen::VectorXd> prev = points;
+	std::vector<Eigen::VectorXd> curr(points.size());
+	bool has_converged = false;
+  int iterations = 0; //for logging
+
+	// The MeanShift loop
+	while (!has_converged) {
```

Comment:
Let's discuss in Skype, I'll let you know when I'm home

Reply:
- On the second point: That is actually a very good point. I have to look at my references again, I might have misinterpreted some variables. Your way acutally makes more sense to me. However it will increase complexity a bit.

- On the third point: I think this will only increase complexity. If two modes are identical (distance < merge threshold) and you don't merge them, you perform the same operations twice in iterate_mode. What would be the advantage with two phases?

- On the fourth point: maybe a simpler solution to the problem would be to use `CONVERGENCE_THRESHOLD*clusters.size()`? This would mean that the average movement is below a certain threshold. Still less accurate than your solution but I think more efficient.